pkg/ring/lifecycler.go

```diff
@@ -375,6 +559,69 @@ heartbeatLoop:
 	}
 }
 
+// waitCleanRing incrementally reads from the KV store and waits
+// until there are no JOINING or LEAVING ingesters.
+func (i *Lifecycler) waitCleanRing(ctx context.Context) error {
+	// Sleep for a random period up to 2s. Used to stagger
```

Comment:
Could you explain? Would we need to set a minimum and maximum stagger? As it is now, it sleeps within the range 0-2s.

Reply:
With gossiping ring, there is a delay between ingester setting its state to JOINING/LEAVING and other ingesters seeing that -- it can easily be 5-10 seconds or even more, depending on size of the cluster and gossiping parameters.

Reply:
Aaah, I've misunderstood what this timeout does. Sorry for confusion.

Reply:
I think you make a good point though, we also want to give new ingesters enough time to see the current view of the ring. I think adding a min and max jitter would help with that.